CORRELATION_CONTEXT_HTTP_HEADER_FORMAT_RATIONALE.md

```diff
@@ -0,0 +1,32 @@
+# Rationale for Correlation Context HTTP Header Format
+
+This document provides rationale for the decisions made for the `Correlation-Context` header format.
+
+## Purpose of a header
+
+Correlation context header is used to propagate properties not defined in `Trace-Context`. There are two common use cases. First is to define a context on trace initiation. Such context will have customer's identity, high-level operation name and other properties like a flight name. Second use case is to pass the caller's name to the next component. This name-value pair will be overridden by every service to it's own name.
+
+## General considerations
+
+- Should be human readable. Cryptic header will hide the fact of potential information disclosure.
+- Should be appende-able (comma-separated) https://tools.ietf.org/html/rfc7230#page-24 so nodes can add context properties without parsing an existing headers.
+- It is expected that the typical name will be a single word in latin and the value will be a short string in latin or a derivative of an url.
+
+## Trimming of spaces
+
+Header should be human readable and editable. Thus spaces are allowed before and after the comma separator.
+
+## Strings encoding
+
+Url encoding is low-overhead way to encode unicode characters for non-latin characters in the values. Url encoding keeps a single words in latin unchanged and easy readable.
+
+## Limits
+
+The idea behind limits is to provide trace vendors common safeguards so the content of the `Correlation-Context` header can be stored with the request. Thus the limits are defined on the number of keys, max pair length and the total size. The last limit is the most important in many scenarios as it allows to plan for the data storage limits.
+
+Another consideration was that cookies may be an alternative way to store and pass the correlation context fields. So the limits should make context name-value pairs fit the typical cookie limit.
+
+- *Maximum number of name-value pairs* - this limit was taken as a number of cookies allowed by Chrome.
+- *Maximum number of bytes per a single name-value pair* - the limit allows to store URL as a value with some extra details as a single context name-value pair. It is also a typical cookie size limitation.
+- *Maximum total length of all name-value pairs* - TODO: LOOKING FOR SUGGESTIONS HERE
+
```

Comment:
tradeoffs:

Another option is to use prefixed headers, such as trace-context-X where X is a propagated field. This could reduce the size of the data, particularly in http/2 where header compression can apply. However XXyour_reason_hereXX

Reply:
another option to add to rational section is http vary style

## Why not Vary-style?
The [Vary](https://tools.ietf.org/html/rfc7231#section-7.1.4) approach is another alternative, which could be used to accomplish the same. For example, `Correlation-Context: x-b3-parentid;ttl=1` could tell the propagation to look at and forward the parent ID header, but only to the next hop. This has an advantage of http header compression (hpack) and also weave-in with legacy tracing headers. However **YOUR_REASON_HERE**

Reply:
> Another option is to use prefixed headers, such as trace-context-X

I'd rather not. Jaeger uses that, and it puts restriction on which keys people want to use for baggage, even when the transport is NOT http but some binary protocol that can perfectly preserve any unicode keys.

Reply:
ps I consider my comment addressed as the rationale section now exists

Reply:
What does "Vary-style" even mean here? Can someone please just specify some [ABNF](https://tools.ietf.org/html/rfc5234) for the syntax being proposed?